### PR TITLE
[Owners] Custom implemention of CreateAdvancedArbSequence API and auto-generation of ErrorHandler API

### PR DESCRIPTION
**Pull request by kruthi-NI**
### Justification

- In the base metadata, `coerced_markers_array` parameter of `CreateAdvancedArbSequence` API is mentioned as `source-code` mechanism. Hence we are custom implementing that API.
- Metadata for `ErrorHandler` API was not added earlier because it was missing in nimi-python metadata and the script we had was not converting it properly from base metadata.
- I had used it for manually testing the `CreateAdvancedArbSequence` API. Hence adding metadata for that API also.

### Implementation

- Added metadata for `ErrorHandler` API and auto-generated it. 
- From C documentation (image attached in below section), I found that the syntax of `ErrorHandler` API was similar to the syntax of `error_message` API, whose metadata was already present in nimi-python [here](https://github.com/ni/nimi-python/blob/master/src/nifgen/metadata/functions.py#L3265). So I used similar pattern to add metadata for  `ErrorHandler` API and verified that it matches the base metadata also.
- Custom implemented `CreateAdvancedArbSequence` API.
- [Link to nimi-python metadata for  `CreateAdvancedArbSequence` API.](https://github.com/ni/nimi-python/blob/master/src/nifgen/metadata/functions.py#L504) 
- [Link to nimi-python helper implementation for `CreateAdvancedArbSequence` API.](https://github.com/ni/nimi-python/blob/master/generated/nifgen/nifgen/session.py#L3437)

### Testing

We have manually tested the API by writing simple Python client application and tested for different parameter cases.
System tests will be added in upcoming PR for the custom API.

### Screenshot of ErrorHandler API from chm used for getting size for metadata:

chm

### CODE REVIEW

File: source/codegen/metadata/nifgen/functions.py [hunk 1]
```diff
@@ -1471,6 +1537,30 @@
         ],
         'returns':'ViStatus'
     },
+    'ErrorHandler':{
```
Comment (kruthi-NI) [inline source/codegen/metadata/nifgen/functions.py:1540]: Note: Attaching screenshot of chm for `ErrorHandler` API to help review the metadata:

chm

File: source/codegen/metadata/nifgen/functions.py [hunk 2]
```diff
@@ -1058,6 +1058,72 @@
...
+                    'value':'sequenceLength'
+                }
+            },
+            {
+                'name':'coercedMarkersArray',
+                'direction':'out',
+                'type':'ViInt32[]',
+                'size':{
+                    'mechanism':'custom-code',
+                    'value':'sequenceLength'
```
Comment (kruthi-NI) [inline source/codegen/metadata/nifgen/functions.py:1116]: @doshirohan , @ganeshnie , in [nimi-python](https://github.com/ni/nimi-python/blob/master/src/nifgen/metadata/functions.py#L584), the value has the following python snippet:
`'value': '(0 if marker_location_array is None else len(marker_location_array))'`.
For gRPC, is it okay to mention as sequenceLength? If not, please suggest what can be used.

File: source/custom/nifgen_service.custom.cpp [hunk 3]
```diff
@@ -1,2 +1,64 @@
+#include <nifgen/nifgen_service.h>
+
+#include <stdexcept>
+
 namespace nifgen_grpc {
-}  // namespace nitclk_grpc
+
+class DriverErrorException : public std::runtime_error{
```
Comment (doshirohan) [inline source/custom/nifgen_service.custom.cpp]: You don't need `CheckStatus` and `DriverErrorException` yet. We only use it if we are making some other library calls before calling the main library API and want to return if those calls return error.

File: source/custom/nifgen_service.custom.cpp [hunk 4]
```diff
@@ -1,2 +1,64 @@
...
+    auto vi_grpc_session = request->vi();
+    ViSession vi = session_repository_->access_session(vi_grpc_session.id(), vi_grpc_session.name());
+    ViInt32 sequence_length = request->waveform_handles_array().size();
+    auto waveform_handles_array = const_cast<ViInt32*>(reinterpret_cast<const ViInt32*>(request->waveform_handles_array().data()));
+    auto loop_counts_array = const_cast<ViInt32*>(reinterpret_cast<const ViInt32*>(request->loop_counts_array().data()));
+    auto sample_counts_array = const_cast<ViInt32*>(reinterpret_cast<const ViInt32*>(request->sample_counts_array().data()));
+    auto marker_location_array = const_cast<ViInt32*>(reinterpret_cast<const ViInt32*>(request->marker_location_array().data()));
+
+    if (marker_location_array != NULL)
+    {
```
Comment (doshirohan) [inline source/custom/nifgen_service.custom.cpp]: Put opening bracket on the line above to keep consistent styling.

**Review by doshirohan — APPROVED** [review]

File: source/codegen/metadata/nifgen/functions.py [hunk 5]
```diff
@@ -1471,6 +1537,30 @@
...
+                'direction':'in',
+                'type':'ViStatus'
+            },
+            {
+                'name':'errorMessage',
+                'direction':'out',
+                'type':'ViChar[]',
+                'size':{
+                    'mechanism':'fixed',
+                    'value':256
```
Comment (Akshata440) [inline source/codegen/metadata/nifgen/functions.py:1474]: You were right about using fixed mechanism and value 256 for this, even in the niFgen.h they have used the same value : 
image

**Review by Akshata440 — APPROVED** [review]

File: source/codegen/metadata/nifgen/functions.py (hunk 5 above)
Comment (kruthi-NI) [reply source/codegen/metadata/nifgen/functions.py:1474]: Oh ok. Thanks Akshata

File: source/custom/nifgen_service.custom.cpp (hunk 3 above)
Comment (kruthi-NI) [reply source/custom/nifgen_service.custom.cpp]: Oh right. Removed it. Thanks Rohan

File: source/custom/nifgen_service.custom.cpp (hunk 4 above)
Comment (kruthi-NI) [reply source/custom/nifgen_service.custom.cpp]: Done

**Review by reckenro — APPROVED** [review]

**Review by jonlugoatx — APPROVED** [review]

**Pull request merged**